src/Rules/Constants/MagicConstantContextRule.php

```diff
@@ -0,0 +1,71 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Constants;
+
+use PhpParser\Node;
+use PhpParser\Node\Scalar\MagicConst;
+use PHPStan\Analyser\Scope;
+use PHPStan\Parser\MagicConstantParamDefaultVisitor;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use function sprintf;
+
+/** @implements Rule<MagicConst> */
+class MagicConstantContextRule implements Rule
+{
+
+	public function getNodeType(): string
+	{
+		return MagicConst::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if ($node instanceof MagicConst\Class_) {
+			if ($scope->isInClass()) {
+				return [];
+			}
+
+			return [
+				RuleErrorBuilder::message(
+					sprintf('Magic constant %s is always empty outside a class.', $node->getName()),
+				)->build(),
+			];
+		} elseif ($node instanceof MagicConst\Trait_) {
+			if ($scope->isInTrait()) {
+				return [];
+			}
+
+			return [
+				RuleErrorBuilder::message(
+					sprintf('Magic constant %s is always empty outside a trait.', $node->getName()),
+				)->build(),
+			];
+		} elseif ($node instanceof MagicConst\Method || $node instanceof MagicConst\Function_) {
+			// https://3v4l.org/3CAHm
```

Comment:
This doesn't seem right. What about `private $foo = __METHOD__;`?

Also a false positive here - you can have an anonymous function outside a class/function and it's going to report `'{closure}'`.